stack/create.go

```diff
@@ -0,0 +1,167 @@
+	Dir string
+
+	// ID of the stack, defaults to an UUID V4 if empty.
+	ID string
+
+	// Name of the stack, defaults to Dir basename if empty.
+	Name string
+
+	// Description of the stack, defaults to Name if empty.
+	Description string
+
+	// Imports represents a set of import paths.
+	Imports []string
+}
+
+// Create creates a stack according to the given configuration.
+// Any dirs on the path provided on the configuration that doesn't exist
+// will be created.
+//
+// If the stack already exists it will return an error and no changes will be
+// made to the stack.
+func Create(rootdir string, cfg CreateCfg) error {
+	const stackFilename = "stack.tm.hcl"
+
+	logger := log.With().
+		Str("action", "stack.Create()").
+		Stringer("cfg", cfg).
+		Logger()
+
+	if !filepath.IsAbs(cfg.Dir) {
+		return errors.E(ErrInvalidStackDir, "relative paths not allowed")
+	}
+
+	if !strings.HasPrefix(cfg.Dir, rootdir) {
+		return errors.E(ErrInvalidStackDir, "stack %q must be inside project root %q", cfg.Dir, rootdir)
+	}
+
+	if strings.HasPrefix(filepath.Base(cfg.Dir), ".") {
+		return errors.E(ErrInvalidStackDir, "dot directories not allowed")
+	}
+
+	logger.Trace().Msg("creating stack dir if absent")
+
+	if err := os.MkdirAll(cfg.Dir, 0755); err != nil {
+		return errors.E(err, "failed to create new stack directories")
+	}
+
+	logger.Trace().Msg("validating create configuration")
+
+	_, err := os.Stat(filepath.Join(cfg.Dir, stackFilename))
+	if err == nil {
+		// Even if there is no stack inside the file, we can't overwrite
+		// the user file anyway.
+		return errors.E(ErrStackAlreadyExists, "%q already exists", stackFilename)
+	}
+
+	// We could have a stack definition somewhere else.
+	parsedCfg, err := hcl.ParseDir(rootdir, cfg.Dir)
+	if err != nil {
+		return errors.E(err, "invalid config creating stack dir %s", cfg.Dir)
+	}
+
+	if parsedCfg.Stack != nil {
+		return errors.E(ErrStackAlreadyExists, "name %q", parsedCfg.Stack.Name)
+	}
+
+	hclCfg, err := hcl.NewConfig(cfg.Dir)
+	if err != nil {
+		return errors.E(err, "failed to create new stack config")
+	}
+
+	if cfg.ID == "" {
+		logger.Trace().Msg("no ID provided, generating one")
+
+		id, err := uuid.NewRandom()
+		if err != nil {
+			return errors.E(err, "failed to create stack UUID")
+		}
+		cfg.ID = id.String()
+	}
+
+	if cfg.Name == "" {
+		cfg.Name = filepath.Base(cfg.Dir)
+	}
+
+	if cfg.Description == "" {
+		cfg.Description = cfg.Name
+	}
+
+	hclID, err := hcl.NewStackID(cfg.ID)
+	if err != nil {
+		return errors.E(ErrInvalidStackID, err)
+	}
+
+	hclCfg.Stack = &hcl.Stack{
+		ID:          hclID,
+		Name:        cfg.Name,
+		Description: cfg.Description,
+	}
+
+	logger.Trace().Msg("creating stack file")
+
+	stackFile, err := os.Create(filepath.Join(cfg.Dir, stackFilename))
+	if err != nil {
+		return errors.E(err, "opening stack file")
+	}
+	defer func() {
+		err := stackFile.Close()
+		if err != nil {
+			logger.Error().Err(err).Msg("closing stack file")
```

Comment:
should not we return an error? because the file can be corrupted (missing the last write).